src/Stratis.Bitcoin/Consensus/PeerGetHeaderRequestBehavior.cs

```diff
@@ -0,0 +1,132 @@
+            IPeerBanning peerBanning)
+        {
+            this.dateTimeProvider = dateTimeProvider;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.peerBanning = peerBanning;
+        }
+
+        public override object Clone()
+        {
+            return new RateLimitingBehavior(this.dateTimeProvider, this.loggerFactory, this.peerBanning);
+        }
+
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceived, true);
+        }
+
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceived);
+        }
+
+        /// <summary>
+        /// Processes an incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private Task OnMessageReceived(INetworkPeer peer, IncomingMessage message)
+        {
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    this.AssessPeerGetHeaderFrequency(getHeaders);
+                    break;
+
+                default:
+                    break;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AssessPeerGetHeaderFrequency(GetHeadersPayload getHeaders)
+        {
+            // Is the last requested hash the same as this request.
+            if (this.getHeaderLastRequestHash == getHeaders.BlockLocator.Blocks.First())
+            {
+                this.logger.LogDebug($"{this.AttachedPeer.PeerEndPoint} block locator matches previous, count {this.getHeaderRequestCount}");
+
+                // Was this hash requested less than 60 seconds ago.
+                if (this.getHeaderLastRequestedTimestamp > this.dateTimeProvider.GetUtcNow().AddSeconds(-60))
+                {
+                    this.getHeaderRequestCount++;
+                }
+                else
+                {
+                    this.getHeaderLastRequestedTimestamp = this.dateTimeProvider.GetUtcNow();
+                    this.getHeaderRequestCount = 0;
+                    return;
+                }
+
+                // If the same header was requested more than 3 times in the last 60 seconds,
+                // ban and disconnect the peer for 1 hour.
+                if (this.getHeaderRequestCount > GetHeaderRequestCountThreshold)
+                {
+                    this.peerBanning.BanAndDisconnectPeer(this.AttachedPeer.PeerEndPoint, BanDurationSeconds, $"Banned via {this.GetType().Name} for {BanDurationSeconds} seconds.");
+                    this.logger.LogDebug($"{this.AttachedPeer.PeerEndPoint} banned via {this.GetType().Name} for {BanDurationSeconds} seconds.");
```

Comment:
Do not use string interpolation in logs please

Reply:
Will Fix